File: packages/engine/src/functions/stream.ts
import { getContext, http } from "iii-sdk";
import type { StateKV } from "../state/kv.js";
import type { EngineConfig } from "../config.js";
import { SCOPES } from "../state/schema.js";
import { getDocker, getContainerStats } from "../docker/client.js";
import { checkAuth } from "../security/validate.js";
import type { Sandbox, SandboxEvent } from "../types.js";

export function registerStreamFunctions(
  sdk: any,
  kv: StateKV,
  config: EngineConfig,
) {
  const getRunningContainer = async (id: string) => {
    const sandbox = await kv.get<Sandbox>(SCOPES.SANDBOXES, id);
    if (!sandbox) throw new Error(`Sandbox not found: ${id}`);
    if (sandbox.status !== "running")
      throw new Error(`Sandbox is not running: ${sandbox.status}`);
    return getDocker().getContainer(`iii-sbx-${id}`);
  };

  sdk.registerFunction(
    { id: "stream::logs", description: "Stream container logs via SSE" },
    http(async (req, res) => {
      const authErr = checkAuth(req as any, config);
      if (authErr) {
        res.status(authErr.status_code);
        res.stream.write(JSON.stringify(authErr.body));
        res.close();
        return;
      }

      const id = req.path_params?.id;
      const follow = req.query_params?.follow !== "false";
      const tail = parseInt(req.query_params?.tail as string, 10) || 100;

      let container;
      try {
        container = await getRunningContainer(id);
      } catch (err: any) {
        const msg = err.message ?? "Internal error";
        const code = msg.includes("not found") ? 404 : 400;
        res.status(code);
        res.stream.write(JSON.stringify({ error: msg }));
        res.close();
        return;
      }

      const ctx = getContext();
      ctx.logger.info("Streaming logs", { id, follow, tail });

      res.status(200);
      res.headers({
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
        Connection: "keep-alive",
      });

      try {
        const logStream = await container.logs({
          follow,
          stdout: true,
          stderr: true,
          tail,
          timestamps: true,
        });

        if (follow && typeof logStream.on === "function") {
          logStream.on("data", (chunk: Buffer) => {
            const lines = chunk.toString().split("\n").filter(Boolean);
            for (const line of lines) {
              const header = line.charCodeAt(0);
              const text = line.length > 8 ? line.slice(8) : line;
              res.stream.write(
                `data: ${JSON.stringify({
                  type: header === 2 ? "stderr" : "stdout",
                  data: text,
                  timestamp: Date.now(),
                })}\n\n`,
              );
            }
          });

          logStream.on("end", () => {
            res.stream.write(
              `data: ${JSON.stringify({ type: "end", data: "", timestamp: Date.now() })}\n\n`,
            );
            res.close();
          });

          logStream.on("error", () => {
            res.close();
          });
        } else {
          const output = typeof logStream === "string" ? logStream : logStream.toString();
          const lines = output.split("\n").filter(Boolean);
          for (const line of lines) {
            const header = line.charCodeAt(0);
            const text = line.length > 8 ? line.slice(8) : line;
            res.stream.write(
              `data: ${JSON.stringify({
                type: header === 2 ? "stderr" : "stdout",
                data: text,
                timestamp: Date.now(),
              })}\n\n`,
            );
          }
          res.stream.write(
            `data: ${JSON.stringify({ type: "end", data: "", timestamp: Date.now() })}\n\n`,
          );
          res.close();
        }
      } catch {
        res.close();
      }
    }),
  );

  sdk.registerFunction(
    { id: "stream::metrics", description: "Stream container metrics via SSE" },
    http(async (req, res) => {
      const authErr = checkAuth(req as any, config);
      if (authErr) {
        res.status(authErr.status_code);
        res.stream.write(JSON.stringify(authErr.body));
        res.close();
        return;
      }

      const id = req.path_params?.id;
      const interval = Math.max(
        parseInt(req.query_params?.interval as string, 10) || 5,
        1,
      );

      let container;
      try {
        container = await getRunningContainer(id);
      } catch (err: any) {
        const msg = err.message ?? "Internal error";
        const code = msg.includes("not found") ? 404 : 400;
        res.status(code);
        res.stream.write(JSON.stringify({ error: msg }));
        res.close();
        return;
      }

      const ctx = getContext();
      ctx.logger.info("Streaming metrics", { id, interval });

      res.status(200);
      res.headers({
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
        Connection: "keep-alive",
      });

      let stopped = false;
      const tick = async () => {
        if (stopped) return;
        try {
          const stats = await getContainerStats(container);
          res.stream.write(`data: ${JSON.stringify(stats)}\n\n`);
        } catch {
          stopped = true;
          res.stream.write(
            `data: ${JSON.stringify({ type: "error", error: "Container stopped", timestamp: Date.now() })}\n\n`,
          );
          res.close();
          return;
        }
        if (!stopped) timer = setTimeout(tick, interval * 1000);
      };

      let timer: ReturnType<typeof setTimeout> = setTimeout(tick, 0);

      req.on?.("close", () => {
        stopped = true;
        clearTimeout(timer);
        res.close();
      });
    }),
  );

  sdk.registerFunction(
    { id: "stream::events", description: "Stream sandbox events via SSE" },
    http(async (req, res) => {
      const authErr = checkAuth(req as any, config);
      if (authErr) {
        res.status(authErr.status_code);
        res.stream.write(JSON.stringify(authErr.body));
        res.close();
        return;
      }

      const topic = req.query_params?.topic as string | undefined;

      const ctx = getContext();
      ctx.logger.info("Streaming events", { topic });

      res.status(200);
      res.headers({
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
        Connection: "keep-alive",
      });

      let lastTimestamp = Date.now();
      let stopped = false;

      const poll = async () => {
        if (stopped) return;
        try {
          let events = await kv.list<SandboxEvent>(SCOPES.EVENTS);
          events = events.filter((e) => e.timestamp > lastTimestamp);
          if (topic) events = events.filter((e) => e.topic === topic);
          events.sort((a, b) => a.timestamp - b.timestamp);

          for (const event of events) {
            res.stream.write(`data: ${JSON.stringify(event)}\n\n`);
            lastTimestamp = event.timestamp;
          }
        } catch {
          stopped = true;
          res.close();
          return;
        }
        if (!stopped) timer = setTimeout(poll, 2000);
      };

      let timer: ReturnType<typeof setTimeout> = setTimeout(poll, 0);

      req.on?.("close", () => {
        stopped = true;
        clearTimeout(timer);
        res.close();
      });
    }),
  );
}

File: packages/worker/src/functions/stream.rs
use bollard::Docker;
use iii_sdk::III;
use serde_json::{json, Value};
use std::sync::Arc;

use crate::config::EngineConfig;
use crate::docker::{container_logs, get_container_stats};
use crate::state::{scopes, StateKV};
use crate::types::{Sandbox, SandboxEvent};

pub fn register(bridge: &Arc<III>, dk: &Arc<Docker>, kv: &StateKV, _config: &EngineConfig) {
    // stream::logs (non-SSE fallback — returns collected log data)
    {
        let kv = kv.clone(); let dk = dk.clone();
        bridge.register_function_with_description("stream::logs", "Stream container logs via SSE", move |input: Value| {
            let kv = kv.clone(); let dk = dk.clone();
            async move {
                let id = input.get("path_params")
                    .and_then(|p| p.get("id"))
                    .and_then(|v| v.as_str())
                    .or_else(|| input.get("id").and_then(|v| v.as_str()))
                    .ok_or_else(|| iii_sdk::IIIError::Handler("id is required".into()))?;

                let sandbox: Sandbox = kv.get(scopes::SANDBOXES, id).await
                    .ok_or_else(|| iii_sdk::IIIError::Handler(format!("Sandbox not found: {id}")))?;
                if sandbox.status != "running" {
                    return Err(iii_sdk::IIIError::Handler(format!("Sandbox is not running: {}", sandbox.status)));
                }

                let query = input.get("query_params").unwrap_or(&Value::Null);
                let tail = query.get("tail").and_then(|v| v.as_str()).unwrap_or("100");

                let cn = format!("iii-sbx-{id}");
                let logs = container_logs(&dk, &cn, false, tail).await
                    .map_err(|e| iii_sdk::IIIError::Handler(e))?;
                Ok(json!({ "logs": logs }))
            }
        });
    }

    // stream::metrics (returns one-shot metrics)
    {
        let kv = kv.clone(); let dk = dk.clone();
        bridge.register_function_with_description("stream::metrics", "Stream resource metrics via SSE", move |input: Value| {
            let kv = kv.clone(); let dk = dk.clone();
            async move {
                let id = input.get("path_params")
                    .and_then(|p| p.get("id"))
                    .and_then(|v| v.as_str())
                    .or_else(|| input.get("id").and_then(|v| v.as_str()))
                    .ok_or_else(|| iii_sdk::IIIError::Handler("id is required".into()))?;

                let sandbox: Sandbox = kv.get(scopes::SANDBOXES, id).await
                    .ok_or_else(|| iii_sdk::IIIError::Handler(format!("Sandbox not found: {id}")))?;
                if sandbox.status != "running" {
                    return Err(iii_sdk::IIIError::Handler(format!("Sandbox is not running: {}", sandbox.status)));
                }

                let cn = format!("iii-sbx-{id}");
                let stats = get_container_stats(&dk, &cn, id).await
                    .map_err(|e| iii_sdk::IIIError::Handler(e))?;
                serde_json::to_value(&stats).map_err(|e| iii_sdk::IIIError::Serde(e.to_string()))
            }
        });
    }

    // stream::events (returns recent events)
    {
        let kv = kv.clone();
        bridge.register_function_with_description("stream::events", "Stream events via SSE", move |input: Value| {
            let kv = kv.clone();
            async move {
                let query = input.get("query_params").unwrap_or(&Value::Null);
                let topic = query.get("topic").and_then(|v| v.as_str());

                let mut events: Vec<SandboxEvent> = kv.list(scopes::EVENTS).await;
                if let Some(t) = topic {
                    events.retain(|e| e.topic == t);
                }
                events.sort_by(|a, b| b.timestamp.cmp(&a.timestamp));
                events.truncate(100);
                Ok(json!({ "events": events }))
            }
        });
    }
}
